packages/contracts/contracts/L1/teleportr/TeleportrDeposit.sol

```diff
@@ -0,0 +1,140 @@
+        uint256 balance = address(this).balance;
+        payable(owner).transfer(balance);
+        emit BalanceWithdrawn(owner, balance);
+    }
+
+    function destroy() public isOwner {
+        emit Destructed(owner, address(this).balance);
+        selfdestruct(payable(owner));
+    }
+
+    // Receive function which reverts if the amount is outside the range
+    // [minDepositAmount, maxDepositAmount], canReceiveDeposit is false, or the
+    // amount would put the contract over its maxBalance.
+    receive()
+        external
+        payable
+        isLowerThanMaxDepositAmount
+        isGreaterThanMinDepositAmount
+        canReceive
+        isLowerThanMaxBalance
+    {
+        uint256 _depositId = totalDeposits + 1;
+        emit EtherReceived(_depositId, msg.sender, msg.value);
+        totalDeposits = _depositId;
+    }
+
+    // Setters
+    function setMaxAmount(uint256 _maxDepositAmount) public isOwner {
+        emit MaxDepositAmountSet(maxDepositAmount, _maxDepositAmount);
+        maxDepositAmount = _maxDepositAmount;
+    }
+
+    function setMinAmount(uint256 _minDepositAmount) public isOwner {
+        emit MinDepositAmountSet(minDepositAmount, _minDepositAmount);
+        minDepositAmount = _minDepositAmount;
+    }
+
+    function setOwner(address newOwner) public isOwner {
+        emit OwnerSet(owner, newOwner);
+        owner = newOwner;
+    }
+
+    function setCanReceiveDeposit(bool _canReceiveDeposit) public isOwner {
+        emit CanReceiveDepositSet(_canReceiveDeposit);
+        canReceiveDeposit = _canReceiveDeposit;
+    }
+
+    function setMaxBalance(uint256 _maxBalance) public isOwner {
+        emit MaxBalanceSet(maxBalance, _maxBalance);
+        maxBalance = _maxBalance;
+    }
+
+    // Getters
+    function getMaxDepositAmount() external view returns (uint256) {
+        return maxDepositAmount;
+    }
+
+    function getMinDepositAmount() external view returns (uint256) {
+        return minDepositAmount;
+    }
+
+    function getMaxBalance() external view returns (uint256) {
+        return maxBalance;
+    }
+
+    function getOwner() external view returns (address) {
+        return owner;
+    }
+
+    function getCanReceiveDeposit() external view returns (bool) {
+        return canReceiveDeposit;
+    }
+
+    // Modifiers
+    modifier isLowerThanMaxDepositAmount() {
+        require(msg.value <= maxDepositAmount, "Deposit amount is too big");
+        _;
+    }
+    modifier isGreaterThanMinDepositAmount() {
+        require(msg.value >= minDepositAmount, "Deposit amount is too small");
+        _;
+    }
+    modifier isOwner() {
+        require(msg.sender == owner, "Caller is not owner");
+        _;
+    }
+    modifier canReceive() {
+        require(canReceiveDeposit, "Contract is not allowed to receive ether");
+        _;
+    }
+    modifier isLowerThanMaxBalance() {
+        require(
+            address(this).balance + msg.value <= maxBalance,
```

Comment:
Oh, this check is broken. 
Ie. `address(this).balance` already includes `msg.value`. So it will revert at a lower balance than it should.

Reply:
ah okay, didn't realize the balance was credited before `receive` was invoked. removed this change.